### PR TITLE
Test fixes for stabilizing tests

### PR DESCRIPTION
**Pull request by adityapatwardhan**
# PR Summary

Make tests more robust when executing on various platforms.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/main/reference/7.3/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: test/tools/NamedPipeConnection/src/code/Microsoft.PowerShell.NamedPipeConnection.csproj [hunk 1]
```diff
@@ -15,8 +15,6 @@
 
   <ItemGroup>
     <PackageReference Include="Microsoft.CSharp" Version="4.7.0" />
-    <Reference Include="System.Management.Automation">
-      <HintPath>.\Ref\System.Management.Automation.dll</HintPath>
-    </Reference>
+    <PackageReference Include="System.Management.Automation" Version="7.4.0-preview.1" />
```
Comment (iSazonov) [inline test/tools/NamedPipeConnection/src/code/Microsoft.PowerShell.NamedPipeConnection.csproj:18]: Should this be automated in some way?

**Comment by adityapatwardhan** [comment]
This PR is under test. Commenting so bot doesnt close the PR.

File: test/tools/NamedPipeConnection/src/code/Microsoft.PowerShell.NamedPipeConnection.csproj (hunk 1 above)
Comment (adityapatwardhan) [reply test/tools/NamedPipeConnection/src/code/Microsoft.PowerShell.NamedPipeConnection.csproj:18]: Good point. I will add this csproj to the dependabot monitored csproj files

File: test/powershell/Host/ConsoleHost.Tests.ps1 [hunk 2]
```diff
@@ -952,8 +952,16 @@ Describe "WindowStyle argument" -Tag Feature {
 
         if ($IsWindows)
         {
-            $ExitCodeBadCommandLineParameter = 64
-            Add-Type -Name User32 -Namespace Test -MemberDefinition @"
+            $sessionId = Get-Process -Id $PID | Select-Object -Property SessionId
```
Comment (adityapatwardhan) [inline test/powershell/Host/ConsoleHost.Tests.ps1:955]: Drop this change

File: test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1 [hunk 3]
```diff
@@ -2094,33 +2095,89 @@ dir -Recurse `
...
+                $aboutHelpPathUserScope = Join-Path $userHelpRoot $culture
+                $aboutHelpPathAllUsersScope = Join-Path $PSHOME $culture
+                $expectedCompletionCount = 0
+
+                ## If help content does not exist, tab completion will not work. So update it first.
+                $userHelpPath = Join-Path $aboutHelpPathUserScope $helpFileName
+                $userScopeHelp = Test-Path $userHelpPath
+                if ($userScopeHelp) {
+                    $expectedCompletionCount++
+                } else {
```
Comment (JamesWTruher) [inline test/powershell/Host/TabCompletion/TabCompletion.Tests.ps1:2148]:
```suggestion
                }
                else {
```

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 4]
```diff
@@ -82,8 +82,12 @@ Describe "Certificate Provider tests" -Tags "Feature" {
     BeforeAll{
         if($IsWindows)
         {
-            Install-TestCertificates
-            Push-Location Cert:\
+            if (-not (Install-TestCertificates) ) {
+                $SetupFailure = $true
+            } else {
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1:87]:
```suggestion
            }
            else {
```

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 5]
```diff
@@ -116,6 +123,9 @@ Describe "Certificate Provider tests" -Tags "Feature" {
         }
         It "Should be able to get DnsNameList of certifate by path: <path>" -TestCases $currentUserMyLocations {
             param([string] $path)
+
+            if ($SetupFailure) { Set-ItResult -Inconclusive -Because "Test certificates are not installed"}
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1:127]:
```suggestion
            if ($SetupFailure) {
                Set-ItResult -Inconclusive -Because "Test certificates are not installed"
            }
```

File: test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1 [hunk 6]
```diff
@@ -133,6 +143,9 @@ Describe "Certificate Provider tests" -Tags "Feature" {
         }
         it "Should be able to get EnhancedKeyUsageList of certifate by path: <path>" -TestCases $currentUserMyLocations {
             param([string] $path)
+
+            if ($SetupFailure) { Set-ItResult -Inconclusive -Because "Test certificates are not installed"}
```
Comment (JamesWTruher) [inline test/powershell/Modules/Microsoft.PowerShell.Security/CertificateProvider.Tests.ps1:147]:
```suggestion
            if ($SetupFailure) {
                Set-ItResult -Inconclusive -Because "Test certificates are not installed"
            }
```

File: test/powershell/engine/COM/COM.Basic.Tests.ps1 [hunk 7]
```diff
@@ -106,12 +106,17 @@ Describe 'Basic COM Tests' -Tags "CI" {
...
 
         It "InvokeMember binder should differentiate PSObject that wraps COM object from other PSObjects" {
-            if (Test-IsWindowsArm64) {
-                Set-ItResult -Pending -Because "COMException: The server process could not be started because the configured identity is incorrect. Check the username and password."
+            try {
+                ## InvokeMember on the member name 'Windows'
+                $shell | ForEach-Object { $_.Windows() } > $null
+            }
+            catch {
+                $notSupported = $_.Exception -is [System.Runtime.InteropServices.COMException]
```
Comment (JamesWTruher) [inline test/powershell/engine/COM/COM.Basic.Tests.ps1:114]: pull this one - it doesn't seem to work

File: test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1 [hunk 8]
```diff
@@ -133,7 +133,7 @@ Describe "Get-ExperimentalFeature Tests" -tags "Feature","RequireAdminOnWindows"
 
 Describe "Default enablement of Experimental Features" -Tags CI {
     BeforeAll {
-        $isPreview = $PSVersionTable.GitCommitId -match "preview|daily"
+        $isPreview = ($PSVersionTable.PSVersion.PreReleaseLabel -ne $null) -or ($PSVersionTable.PSVersion.PreReleaseLabel -notmatch "^RC")
```
Comment (JamesWTruher) [inline test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1:136]: is there a common function we can use?

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1 [hunk 9]
```diff
@@ -61,7 +61,7 @@ Describe "JEA session Transcript script test" -Tag @("Feature", 'RequireAdminOnW
     BeforeAll {
         $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
 
-        if ( ! $IsWindows -or !(Test-CanWriteToPsHome))
+        if ( ! $IsWindows -or !(Test-CanWriteToPsHome) -or (Test-IsWinWow64))
         {
             $PSDefaultParameterValues["it:skip"] = $true
```
Comment (JamesWTruher) [inline test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1:66]:
```suggestion
            $global:PSDefaultParameterValues["it:skip"] = $true
```

File: test/powershell/Host/ConsoleHost.Tests.ps1 (hunk 2 above)
Comment (adityapatwardhan) [reply test/powershell/Host/ConsoleHost.Tests.ps1:955]: Removed

File: test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1 (hunk 9 above)
Comment (adityapatwardhan) [reply test/powershell/engine/Remoting/RemoteSession.Basic.Tests.ps1:66]: Reverting this file

File: test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1 (hunk 8 above)
Comment (adityapatwardhan) [reply test/powershell/engine/ExperimentalFeature/Get-ExperimentalFeature.Tests.ps1:136]: Moved to common function and used here

**Comment by SavageHaxCash** [comment]
can you tell me why im a noob Sent from my iPhoneOn Mar 10, 2023, at 1:59 PM, Aditya Patwardhan ***@***.***> wrote:﻿
@adityapatwardhan commented on this pull request.



In test/powershell/Host/ConsoleHost.Tests.ps1:
> @@ -952,8 +952,16 @@ Describe "WindowStyle argument" -Tag Feature {
 
         if ($IsWindows)
         {
-            $ExitCodeBadCommandLineParameter = 64
-            Add-Type -Name User32 -Namespace Test -MemberDefinition @"
+            $sessionId = Get-Process -Id $PID | Select-Object -Property SessionId

Drop this change

**Comment by adityapatwardhan** [comment]
@SavageHaxCash this change was added for debugging the test failure. It did not give us enough information to identify and fix the problem. We would be handling this issue in later PRs.

File: test/powershell/Host/HostUtilities.Tests.ps1 [hunk 10]
```diff
@@ -35,6 +35,12 @@ Describe "InvokeOnRunspace method as nested command" -tags "Feature" {
 Describe "InvokeOnRunspace method on remote runspace" -tags "Feature","RequireAdminOnWindows" {
 
     BeforeAll {
+        $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
+
+        if (Test-IsWinWow64) {
```
Comment (TravisEz13) [inline test/powershell/Host/HostUtilities.Tests.ps1:40]: add a comment on why we are skipping

File: tools/releaseBuild/azureDevOps/templates/linux-authenticode-sign.yml [hunk 11]
```diff
@@ -2,6 +2,7 @@ jobs:
 - job: sign_linux_builds
   displayName: Sign all linux builds
   condition: succeeded()
+  timeoutInMinutes: 90
```
Comment (TravisEz13) [inline tools/releaseBuild/azureDevOps/templates/linux-authenticode-sign.yml:5]: should this be in a separate PR for backporting?

Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/templates/linux-authenticode-sign.yml:5]: yes.. i will revert this change.

**Comment by TravisEz13** [comment]
We should avoid checking in binary files.  Some other options to consider:

- build one as a test tool
- put one in a blob
- put it in another repo (preferably with [LFS](https://docs.github.com/en/repositories/working-with-files/managing-large-files/about-git-large-file-storage) enabled)
- put it in a feed

File: tools/releaseBuild/azureDevOps/templates/linux-authenticode-sign.yml (hunk 11 above)
Comment (adityapatwardhan) [reply tools/releaseBuild/azureDevOps/templates/linux-authenticode-sign.yml:5]: Reverted

**Comment by adityapatwardhan** [comment]
Removed the change adding the asset and also the change setting timeout for the pipeline task

File: test/powershell/Host/HostUtilities.Tests.ps1 (hunk 10 above)
Comment (adityapatwardhan) [reply test/powershell/Host/HostUtilities.Tests.ps1:40]: We decided to skip remoting test on wow64 as it is too difficult to configure and we dont know if there is enough or any usage for this scenario

**Review by TravisEz13 — APPROVED** [review]
